### PR TITLE
DEV - Update dev workflows and CI

### PR DESCRIPTION
**Pull request by trallard**
Split from #792 

## Description

- Remove OS-specific `environment.yaml` files as no longer needed
- Update env files - server
- Update environment file - core
- Add note about updating dependencies
- Update CI - remove unnecessary `yaml` files

## Pull request checklist

- [x] Did you test this change locally?
- [x] Did you update the documentation (if required)?
- [ ] Did you add/update relevant tests for this change (if required)?

## Additional information

With this PR, at most, contributors/maintainers will need to regenerate/update their local conda environment. 

## How to test

💻 Locally:

1. Regenerate your local `conda` development environment (should use `conda-store-server/environment-dev.yaml`
2. Perform your usual development tasks such as running tests, building conda-store, etc.

To test the changes to the test GH workflows, you can use https://github.com/nektos/act and run the workflow or job accordingly. For example, to only run the conda-store integration tests: `act -j "integration-test-conda-store-server"`

### CODE REVIEW

File: .github/workflows/tests.yaml [hunk 1]
```diff
@@ -22,13 +20,6 @@ jobs:
     strategy:
       matrix:
         os: ["ubuntu", "macos", "windows"]
```
Comment (jaimergp) [inline .github/workflows/tests.yaml:22]:
```suggestion
        os: ["ubuntu-latest", "macos-latest", "macos-14", "windows-latest"]
```

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -22,13 +20,6 @@ jobs:
...
-        include:
-          - os: ubuntu
-            environment-file: conda-store-server/environment-dev.yaml
-          - os: macos
-            environment-file: conda-store-server/environment-macos-dev.yaml
-          - os: windows
-            environment-file: conda-store-server/environment-windows-dev.yaml
     runs-on: ${{ matrix.os }}-latest
```
Comment (jaimergp) [inline .github/workflows/tests.yaml:23]:
```suggestion
    runs-on: ${{ matrix.os }}
```

File: conda-store-server/environment-dev.yaml [hunk 3]
```diff
@@ -1,63 +1,31 @@
...
+# required conda-store-server or conda-store (core) dependencies MUST ONLY be added/updated in the [dependencies]
+# section of the corresponding pyproject.toml file
+# the only exceptions to this rule are conda and constructor which are only conda installable
 name: conda-store-server-dev
 channels:
   - conda-forge
-  - microsoft
   - nodefaults
 dependencies:
   - python ==3.10
```
Comment (jaimergp) [inline conda-store-server/environment-dev.yaml:11]:
```suggestion
  - python =3.10
```

`==` is too strict. `=` allows `3.10.*` updates.

File: conda-store-server/environment-dev.yaml [hunk 4]
```diff
@@ -1,63 +1,31 @@
...
   # conda builds
   - conda ==23.5.2
```
Comment (jaimergp) [inline conda-store-server/environment-dev.yaml:13]: Why this particular pin?

File: conda-store-server/environment-dev.yaml [hunk 5]
```diff
@@ -1,63 +1,31 @@
...
+  - build
+  # installer
+  - constructor # runtime dep but must be conda installed
   - pip
-
   - pip:
       - pytest-playwright
+      - pre-commit
+      - twine>=5.0.0
+      - pkginfo>=1.10 # needed to support metadata 2.3
```
Comment (jaimergp) [inline conda-store-server/environment-dev.yaml:31]: Can these be conda deps?

File: conda-store-server/environment.yaml
Comment (jaimergp) [inline conda-store-server/environment.yaml:1]: What's this environment file for?

File: conda-store-server/pyproject.toml [hunk 6]
```diff
@@ -68,21 +68,20 @@ Source = "https://github.com/conda-incubator/conda-store"
 [tool.hatch.version]
 path = "conda_store_server/__init__.py"
 
+# when adding dev dependencies (not required conda-store-server) add them to the environment-dev.yaml file too
```
Comment (jaimergp) [inline conda-store-server/pyproject.toml:71]: I am afraid that without a CI step that checks for this cross-file consistency, things will go out of sync eventually.

File: conda-store/environment.yaml [hunk 7]
```diff
@@ -2,19 +2,14 @@ name: conda-store
 channels:
   - conda-forge
 dependencies:
-  - pip
   - jupyterhub<2.0.0
-  - jupyterlab>=4.0.0
+  - jupyterlab>4.0
```
Comment (jaimergp) [inline conda-store/environment.yaml:6]:
```suggestion
  - jupyterlab>=4.0
```

File: conda-store/environment.yaml [hunk 8]
```diff
@@ -2,19 +2,14 @@ name: conda-store
...
-  - yarl
-  - aiohttp>=3.8.1
-  - rich
-  - click
-  - ruamel.yaml
+  # installer
+  - constructor # runtime dep but must be conda installed
+  - pip
   - pip:
       - https://github.com/yuvipanda/jupyter-launcher-shortcuts/archive/refs/heads/master.zip
```
Comment (jaimergp) [inline conda-store/environment.yaml:14]: This should be _at least_ a commit hash. But ideally we switch to a released (conda) package.

File: docusaurus-docs/community/contribute/contribute-code.md [hunk 9]
```diff
@@ -5,32 +5,50 @@ description: Contribute to conda-store's codebase
...
 there will be a performance impact when building and running on
-arm architectures. Otherwise, this workflow has been shown to run and build on OSX.
-Notice the `architecture: amd64` within the `docker-compose.yaml` files.
+arm architectures.
+Otherwise, this workflow has been shown to run and build on OSX.
+**Notice** the `architecture: amd64` within the `docker-compose.yaml` files.
 :::
 
-The following resources will be available:
+After running the docker-compose command, the following resources will be available:
```
Comment (jaimergp) [inline docusaurus-docs/community/contribute/contribute-code.md:51]:
```suggestion
After running the `docker-compose` command, the following resources will be available:
```

File: .github/workflows/tests.yaml (hunk 1 above)
Comment (trallard) [reply .github/workflows/tests.yaml:22]: yay! yes this is what I was asking on your PR

File: conda-store-server/environment-dev.yaml (hunk 3 above)
Comment (trallard) [reply conda-store-server/environment-dev.yaml:11]: I had missed we had a `==` 👀  going blind

File: conda-store-server/environment-dev.yaml (hunk 4 above)
Comment (trallard) [reply conda-store-server/environment-dev.yaml:13]: Seems this was introduced in https://github.com/conda-incubator/conda-store/pull/522 -> going through the commits this seems was temporary while https://github.com/mamba-org/mamba/issues/2715

Will remove it and see how it goes

File: conda-store-server/environment-dev.yaml (hunk 5 above)
Comment (trallard) [reply conda-store-server/environment-dev.yaml:31]: Consider it done

File: conda-store-server/environment.yaml
Comment (trallard) [reply conda-store-server/environment.yaml:1]: This is used with docker-compose - will add a note at the top

File: conda-store-server/pyproject.toml (hunk 6 above)
Comment (trallard) [reply conda-store-server/pyproject.toml:71]: +1 - will have to add that in a separate PR (already have other CI changes that build on this PR to simplify stuff and add consistency checks)

File: conda-store/environment.yaml (hunk 8 above)
Comment (trallard) [reply conda-store/environment.yaml:14]: I do not know why this is even here 🤷🏽 will replace with the released package and see how it goes

**Review by trallard — COMMENTED** [review]
Thank you for the quick and thorough review @jaimergp

File: .github/workflows/tests.yaml [hunk 10]
```diff
@@ -21,15 +20,8 @@ jobs:
     name: "unit-test conda-store-server"
     strategy:
```
Comment (jaimergp) [inline .github/workflows/tests.yaml:21]: Can we (temporarily?) add this to ensure that all CI entries run well in this PR?

```suggestion
    strategy:
      fail-fast: false
```

File: .github/workflows/tests.yaml [hunk 11]
```diff
@@ -1,16 +1,15 @@
...
 on:
   pull_request:
   push:
     branches:
       - main
 
-
 # ensuring only one instance is running at a given time
 concurrency:
   group: ${{ github.ref }}
```
Comment (jaimergp) [inline .github/workflows/tests.yaml:15]:
```suggestion
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.sha }}
```

[Explanation in `conda/conda`](https://github.com/conda/conda/blob/11c21d523f2e18e55b07479a31681cd4b63b8517/.github/workflows/tests.yml#L18-L25)

File: .github/workflows/tests.yaml [hunk 12]
```diff
@@ -41,7 +33,7 @@ jobs:
       - name: "Set up env ${{ matrix.os }} 🐍"
```
Comment (jaimergp) [inline .github/workflows/tests.yaml:33]: The sibling action in the other job uses a different name without the platform. Might be good to homogeneize.

File: docusaurus-docs/community/contribute/contribute-code.md [hunk 13]
```diff
@@ -40,7 +58,7 @@ The following resources will be available:
 | [PostgreSQL](https://www.postgresql.org/) (database: `conda-store`)| [localhost:5432](http://localhost:5432) | `admin` | `password` |
 | [Redis](https://www.redis.com/) |  [localhost:6379](http://localhost:6379) | - | password |
 
-On a fast machine this deployment should only take 10 or so seconds
+On a fast machine, this deployment should only take 10 or so seconds
 assuming the docker images have been partially built before.
```
Comment (jaimergp) [inline docusaurus-docs/community/contribute/contribute-code.md:62]:
```suggestion
assuming the Docker images have been partially built before.
```

File: .github/workflows/tests.yaml (hunk 12 above)
Comment (trallard) [reply .github/workflows/tests.yaml:33]: Which sibling action?

Comment (jaimergp) [reply .github/workflows/tests.yaml:34]: The other setup-miniconda instance in this workflow

File: conda-store/pyproject.toml [hunk 14]
```diff
@@ -48,14 +48,16 @@ path = "conda_store/__init__.py"
 [project.optional-dependencies]
 dev = ["build", "twine"]
 
+# when adding dev dependencies (not required conda-store dependencies)
+# you must make sute to add them to the environment-dev.yaml file too
```
Comment (peytondmurray) [inline conda-store/pyproject.toml:52]: :+1:

File: docusaurus-docs/community/contribute/contribute-code.md [hunk 15]
```diff
@@ -126,23 +215,46 @@ Open your local browser and go to [http://localhost:8000](http://localhost:8000)
...
-`c.CondaStoreServer.url_prefix` is set to `/` when generating the
-endpoints.
-<!-- TODO -->
+The REST API is considered somewhat stable. If any changes are made to the API make sure the update the OpenAPI/Swagger
+specification in `docs/_static/openapi.json`.
+This may be downloaded from the `/docs` endpoint when running conda-store.
+Ensure that the `c.CondaStoreServer.url_prefix` is set to `/` when generating the endpoints.
 
-## Workflows
+### Adding new dependencies to the libraries
```
Comment (peytondmurray) [inline docusaurus-docs/community/contribute/contribute-code.md:225]: Thanks for this!

File: .github/workflows/tests.yaml (hunk 12 above)
Comment (trallard) [reply .github/workflows/tests.yaml:34]: ah, added the name manually

**Pull request closed without merging**